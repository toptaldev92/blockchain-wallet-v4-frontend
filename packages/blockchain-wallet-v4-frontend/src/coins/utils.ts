import { ADDRESS_TYPES } from 'core/redux/payment/btc/utils'
import { CoinType, Erc20CoinsEnum, SBBalanceType } from 'core/types'

export const generateCustodyAccount = (coin: CoinType, sbBalance?: SBBalanceType) => {
  // hack to support PAX rebrand ðŸ¤¬
  const ticker = coin === 'PAX' ? 'USD-D' : coin
  return [
    {
      baseCoin: coin in Erc20CoinsEnum ? 'ETH' : coin,
      coin,
      label: `${ticker} Trading Wallet`,
      type: ADDRESS_TYPES.CUSTODIAL,
      balance: sbBalance?.available || '0'
    }
  ]
}