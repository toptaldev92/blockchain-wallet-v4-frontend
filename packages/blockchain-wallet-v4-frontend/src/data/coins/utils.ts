import {
  CoinType,
  InterestBalanceType,
  SBBalanceType,
  SupportedCoinType
} from 'blockchain-wallet-v4/src/types'
import { convertStandardToBase } from 'data/components/exchange/services'
import { SwapAccountType, SwapBaseCounterTypes } from 'data/components/types'

export const generateTradingAccount = (
  coin: CoinType,
  config: SupportedCoinType,
  sbBalance?: SBBalanceType
): SwapAccountType[] => {
  return [
    {
      balance: sbBalance?.available || '0',
      baseCoin: config.contractAddress ? 'ETH' : (coin as SwapAccountType['baseCoin']),
      coin,
      config,
      label: 'Trading Account',
      type: SwapBaseCounterTypes.CUSTODIAL
    }
  ]
}

export const generateInterestAccount = (
  coin: CoinType,
  config: SupportedCoinType,
  interestBalance?: InterestBalanceType
): SwapAccountType[] => {
  return [
    {
      balance: interestBalance?.balance || '0',
      baseCoin: config.contractAddress ? 'ETH' : (coin as SwapAccountType['baseCoin']),
      coin,
      config,
      label: `${coin} Interest Account`,
      // @ts-ignore
      type: 'INTEREST'
    }
  ]
}

export const generateProvisionalPaymentAmount = (
  coin: CoinType,
  amount: number
): string | number => {
  if (coin === 'BTC' || coin === 'BCH') {
    return parseInt(convertStandardToBase(coin, amount))
  }

  return convertStandardToBase(coin, amount)
}
