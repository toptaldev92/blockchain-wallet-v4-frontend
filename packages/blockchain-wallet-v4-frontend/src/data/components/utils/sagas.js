import { select } from 'redux-saga/effects'
import { concat, is, isNil, prop } from 'ramda'
import { selectors } from 'data'
import settings from 'config'

export const selectAccounts = function * () {
  const btcHDAccountsInfo = yield select(selectors.core.common.bitcoin.getAccountsInfo)
  const btcAddressesInfo = yield select(selectors.core.common.bitcoin.getAddressesInfo)
  const btcAccountsInfo = concat(btcHDAccountsInfo, btcAddressesInfo)
  const ethAccountsInfoR = yield select(selectors.core.common.ethereum.getAccountsInfo)
  const ethAccountsInfo = ethAccountsInfoR.getOrElse([])
  return {
    btcAccountsInfo,
    ethAccountsInfo
  }
}

export const selectFee = function * (coin) {
  // const bchRatesR = yield select(selectors2.data.bch.getFee)
  const btcRatesR = yield select(selectors.core.data.btc.getFee)
  const ethRatesR = yield select(selectors.core.data.eth.getFee)
  switch (coin) {
    // case 'BCH': return prop('priority', bchRatesR.getOrElse({ regular: 2, priority: 10 }))
    case 'BTC': return btcRatesR.getOrElse({ regular: 10, priority: 25 })
    case 'ETH': return ethRatesR.getOrElse({ regular: 21, priority: 21, gasLimit: 23000 })
  }
}

export const selectRates = function * (coin) {
  // const bchRatesR = yield select(selectors2.data.bch.getRates)
  const btcRatesR = yield select(selectors.core.data.btc.getRates)
  const ethRatesR = yield select(selectors.core.data.eth.getRates)
  switch (coin) {
    // case 'BCH': return bchRatesR.getOrElse({})
    case 'BTC': return btcRatesR.getOrElse({})
    case 'ETH': return ethRatesR.getOrElse({})
  }
}

export const selectShapeshiftPair = function * (pair) {
  const shapeshiftPairR = yield select(selectors.core.data.shapeshift.getPair(pair))
  const shapeshiftPair = shapeshiftPairR.getOrElse({})
  return {
    minimum: prop('minimum', shapeshiftPair),
    maximum: prop('limit', shapeshiftPair)
  }
}

export const selectReceiveAddress = function * (source) {
  const address = prop('address', source)
  const index = prop('index', source)
  if (!isNil(address) && is(String, address)) {
    return address
  }
  if (!isNil(index) && is(Number, index)) {
    const addressR = yield select(selectors.core.common.bitcoin.getNextAvailableReceiveAddress(settings.NETWORK_BITCOIN, index))
    return addressR.getOrElse('')
  }
  throw new Error('Could not generate next BTC receive address')
}

export const selectChangeAddress = function * (source) {
  const address = prop('address', source)
  const index = prop('index', source)
  if (!isNil(address) && is(String, index)) {
    return address
  }
  if (!isNil(index) && is(Number, index)) {
    const addressR = yield select(selectors.core.common.bitcoin.getNextAvailableChangeAddress(settings.NETWORK_BITCOIN, index))
    return addressR.getOrElse('')
  }
  throw new Error('Could not generate next BTC change address')
}
