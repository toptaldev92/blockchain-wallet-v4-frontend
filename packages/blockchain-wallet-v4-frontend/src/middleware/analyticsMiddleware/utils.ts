import crypto from 'crypto'
import type {
  BuySellClickedOrigin,
  InterestDepositClickedOrigin
} from 'middleware/analyticsMiddleware/types'
import { PaymentType } from 'middleware/analyticsMiddleware/types'

import { PaymentValue, SBPaymentTypes } from 'blockchain-wallet-v4/src/types'
import { SBShowModalOriginType } from 'data/types'

const buySellClickedOriginDictionary = (
  rawOrigin: SBShowModalOriginType | string
): BuySellClickedOrigin => {
  switch (rawOrigin) {
    case 'InterestPage':
      return 'SAVINGS'
    case 'PendingOrder':
      return 'PENDING_ORDER'
    case 'SideNav':
      return 'NAVIGATION'
    case 'WelcomeModal':
      return 'WELCOME'
    case 'WithdrawModal':
      return 'LINK_BANK'
    case 'PriceChart':
      return 'PRICE_CHART'
    case 'GOALS':
    case 'SimpleBuyLink':
      return 'BUY_WIDGET'
    case 'CurrencyList':
      return 'CURRENCY_PAGE'
    case 'TransactionList':
      return 'TRANSACTION_LIST'
    case 'SettingsProfile':
    case 'SettingsGeneral':
      return 'SETTINGS'
    case 'EmptyFeed':
    case 'SwapNoHoldings':
      return 'EMPTY_FEED'
    case 'Goals':
      return 'DEEP_LINK'
    default: {
      throw new Error('Origin not found')
    }
  }
}

const buyPaymentMethodSelectedPaymentTypeDictionary = (
  rawPaymentType: SBPaymentTypes
): PaymentType => {
  switch (rawPaymentType) {
    case SBPaymentTypes.USER_CARD: {
      return PaymentType.PAYMENT_CARD
    }
    case SBPaymentTypes.LINK_BANK: {
      return PaymentType.BANK_ACCOUNT
    }
    case SBPaymentTypes.BANK_ACCOUNT: {
      return PaymentType.BANK_ACCOUNT
    }
    case SBPaymentTypes.FUNDS: {
      return PaymentType.FUNDS
    }
    case SBPaymentTypes.BANK_TRANSFER: {
      return PaymentType.BANK_TRANSFER
    }
    default: {
      return PaymentType.BANK_ACCOUNT
    }
  }
}

const getOriginalTimestamp = () => new Date().toISOString()

const generateUniqueUserId = (guid: string) => {
  return crypto.createHash('sha256').update(guid).digest().toString('base64')
}

const getNetworkFee = (paymentValue: PaymentValue | null) => {
  return paymentValue
    ? paymentValue.coin === 'BTC' || paymentValue.coin === 'BCH'
      ? paymentValue.selection?.fee
      : paymentValue.fee
    : 0
}

const interestDepositClickedOriginDictionary = (rawOrigin): InterestDepositClickedOrigin => {
  switch (rawOrigin) {
    case 'InterestPage':
      return 'SAVINGS_PAGE'
    default: {
      throw new Error('Origin not found')
    }
  }
}

export {
  buyPaymentMethodSelectedPaymentTypeDictionary,
  buySellClickedOriginDictionary,
  generateUniqueUserId,
  getNetworkFee,
  getOriginalTimestamp,
  interestDepositClickedOriginDictionary
}
