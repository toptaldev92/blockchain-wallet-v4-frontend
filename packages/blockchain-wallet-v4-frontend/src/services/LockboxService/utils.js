import { Observable } from 'rxjs'
import React from 'react'
import { FormattedMessage } from 'react-intl'

import firmware from './firmware'
import constants from './constants'

/**
 * Creates device socket
 * @param {Transport} transport - Current device transport
 * @param {String} url - The web socket url to connect to
 * @returns {Observable} the final socket result
 */
/* eslint-disable */
const createDeviceSocket = (transport, url) => {
  return Observable.create(o => {
    let ws, lastMessage

    try {
      ws = new WebSocket(url)
    } catch (err) {
      o.error(err.message, { url })
      return () => {}
    }

    ws.onopen = () => {
      console.info('OPENED', { url })
    }

    ws.onerror = e => {
      console.info('ERROR', { message: e.message, stack: e.stack })
      o.error(e.message, { url })
    }

    ws.onclose = () => {
      console.info('CLOSE')
      o.next(lastMessage || '')
      o.complete()
    }

    ws.onmessage = async rawMsg => {
      try {
        const msg = JSON.parse(rawMsg.data)
        if (!(msg.query in handlers)) {
          throw new Error({ message: 's0ck3t' }, { url })
        }
        console.info('RECEIVE', msg)
        await handlers[msg.query](msg)
      } catch (err) {
        console.info('ERROR', { message: err.message, stack: err.stack })
        o.error(err)
      }
    }

    const send = (nonce, response, data) => {
      const msg = {
        nonce,
        response,
        data
      }
      console.info('SEND', msg)
      const strMsg = JSON.stringify(msg)
      ws.send(strMsg)
    }

    const handlers = {
      exchange: async input => {
        const { data, nonce } = input
        const res = await transport.exchange(Buffer.from(data, 'hex'))
        const status = res.slice(res.length - 2)
        const buffer = res.slice(0, res.length - 2)
        const strStatus = status.toString('hex')
        send(
          nonce,
          strStatus === '9000' ? 'success' : 'error',
          buffer.toString('hex')
        )
      },

      bulk: async input => {
        const { data, nonce } = input
        let lastStatus // Execute all apdus and collect last status

        for (const apdu of data) {
          const res = await transport.exchange(Buffer.from(apdu, 'hex'))
          lastStatus = res.slice(res.length - 2)

          if (lastStatus.toString('hex') !== '9000') break
        }

        if (!lastStatus) {
          throw new Error({ message: 's0ck3t' }, { url })
        }

        const strStatus = lastStatus.toString('hex')

        send(
          nonce,
          strStatus === '9000' ? 'success' : 'error',
          strStatus === '9000' ? '' : strStatus
        )
      },

      success: msg => {
        lastMessage = msg.data || msg.result
        ws.close()
      },

      error: msg => {
        debugger
        console.info('ERROR', { data: msg.data })
        ws.close()
        throw new Error(msg.data, { url })
      }
    }

    return () => {
      if (ws.readyState === 1) {
        lastMessage = null
        ws.close()
      }
    }
  })
}
/* eslint-enable */

/**
 * Gets and parses full device information from api response
 * @param {Transport} transport - Current device transport
 * @returns {Promise} full device information
 */
const getDeviceInfo = transport => {
  return new Promise((resolve, reject) => {
    firmware.getDeviceFirmwareInfo(transport).then(
      res => {
        const { seVersion } = res
        const { targetId, mcuVersion, flags } = res
        const parsedVersion =
          seVersion.match(
            /([0-9]+.[0-9])+(.[0-9]+)?((?!-osu)-([a-z]+))?(-osu)?/
          ) || []
        const isOSU = typeof parsedVersion[5] !== 'undefined'
        const providerName = parsedVersion[4] || ''
        const providerId = constants.providers[providerName]
        const isBootloader = targetId === 0x01000001
        const majMin = parsedVersion[1]
        const patch = parsedVersion[2] || '.0'
        const fullVersion = `${majMin}${patch}${
          providerName ? `-${providerName}` : ''
        }`
        resolve({
          targetId,
          seVersion: majMin + patch,
          isOSU,
          mcuVersion,
          isBootloader,
          providerName,
          providerId,
          flags,
          fullVersion
        })
      },
      err => {
        reject(err)
      }
    )
  })
}

/**
 * Maps a socket error code to a human readable error
 * @param {Promise} promise - Current device transport
 * @returns {Promise} a catch function that returns human error
 */
const mapSocketError = promise => {
  return promise.catch(err => {
    switch (true) {
      case err.message.endsWith('6985'):
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.connectionrefused'
              defaultMessage='Device connection was refused'
            />
          )
        }
      case err.message.endsWith('6982'):
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.devicelocked'
              defaultMessage='Device locked and unable to communicate'
            />
          )
        }
      case err.message.endsWith('6a84') || err.message.endsWith('6a85'):
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.storagespace'
              defaultMessage='Insufficient storage space on device'
            />
          )
        }
      case err.message.endsWith('6a80') || err.message.endsWith('6a81'):
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.appalreadyinstalled'
              defaultMessage='App already installed'
            />
          )
        }
      case err.message.endsWith('6a83'):
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.btcapprequired'
              defaultMessage='Unable to remove BTC app as it is required by others'
            />
          )
        }
      case err.message.endsWith('s0ck3t'):
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.socketerror'
              defaultMessage='Socket connection failed'
            />
          )
        }
      default:
        return {
          err,
          errMsg: () => (
            <FormattedMessage
              id='lockbox.service.messages.unknownerror'
              defaultMessage='An unknown error has occurred'
            />
          )
        }
    }
  })
}

/**
 * Determines correct scrambleKey to use for device connection
 * @param {String} app - Current app requested
 * @param {String} deviceType - Either 'ledger' or 'blockchain'
 * @returns {String} the scrambleKey for connection
 */
const getScrambleKey = (app, deviceType) => {
  return constants.scrambleKeys[deviceType][app]
}

export default {
  createDeviceSocket,
  getDeviceInfo,
  getScrambleKey,
  mapSocketError
}
