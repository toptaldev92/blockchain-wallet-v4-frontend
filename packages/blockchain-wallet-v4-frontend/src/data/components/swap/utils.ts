import { errorHandler } from 'blockchain-wallet-v4/src/utils'
import { SwapQuoteType } from 'core/types'
import BigNumber from 'bignumber.js'

export const getRate = (
  priceTiers: SwapQuoteType['quote']['priceTiers'],
  amount: BigNumber
): number => {
  try {
    for (var index = 0; index <= priceTiers.length; index++) {
      const priceTier = priceTiers[index]
      if (index === priceTiers.length - 1) return Number(priceTier.price)

      const nextTier = priceTiers[index + 1]
      const thisVol = new BigNumber(priceTier.volume)
      const nextVol = new BigNumber(nextTier.volume)

      if (thisVol.isLessThan(amount) && amount.isLessThanOrEqualTo(nextVol)) {
        const price = interpolatePrice(
          new BigNumber(priceTier.volume),
          new BigNumber(priceTier.price),
          new BigNumber(nextTier.volume),
          new BigNumber(nextTier.price),
          amount
        )

        if (typeof price === 'string') throw price
        return price
      }
    }

    return 0
  } catch (e) {
    throw Error(errorHandler(e))
  }
}

export const interpolatePrice = (
  thisVol: BigNumber,
  thisPrice: BigNumber,
  nextVol: BigNumber,
  nextPrice: BigNumber,
  amount: BigNumber
): number | string => {
  try {
    if (thisPrice.isGreaterThan(nextPrice)) throw new Error('Should be sorted')
    if (thisVol.isGreaterThan(nextVol)) throw new Error('Should be sorted')
    if (amount.isLessThan(thisVol) && amount.isGreaterThan(nextVol))
      throw new Error('Amount should be between x[0] and x[1]')

    return amount
      .minus(thisVol)
      .times(nextPrice.minus(thisPrice).dividedBy(nextVol.minus(thisVol)))
      .plus(thisPrice)
      .toNumber()
  } catch (e) {
    return errorHandler(e)
  }
}
