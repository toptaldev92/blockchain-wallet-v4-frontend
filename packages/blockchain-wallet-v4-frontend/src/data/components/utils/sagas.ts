import { equals, identity, includes, is, isEmpty, prop } from 'ramda'
import { select } from 'redux-saga/effects'
import EthUtil from 'ethereumjs-util'

import { ADDRESS_TYPES } from 'blockchain-wallet-v4/src/redux/payment/btc/utils'
import { selectors } from 'data'
import { utils } from 'blockchain-wallet-v4/src'

export const selectReceiveAddress = function * (source, networks) {
  const appState = yield select(identity)
  const erc20List = (yield select(
    selectors.core.walletOptions.getErc20CoinList
  )).getOrFail()
  const coin = prop('coin', source)
  const type = prop('type', source)
  const address = prop('address', source)
  if (equals('XLM', coin) && is(String, address)) return address
  if (
    (includes(coin, erc20List) || equals('ETH', coin)) &&
    is(String, address)
  ) {
    return EthUtil.toChecksumAddress(address)
  }
  if (equals('BCH', coin)) {
    const selector =
      type !== ADDRESS_TYPES.LOCKBOX
        ? selectors.core.common.bch.getNextAvailableReceiveAddress
        : selectors.core.common.bch.getNextAvailableReceiveAddressLockbox
    const bchReceiveAddress = selector(networks.bch, address, appState)
    if (isEmpty(bchReceiveAddress.getOrElse(''))) {
      throw new Error('Could not generate bitcoin cash receive address')
    }
    return utils.bch.toCashAddr(bchReceiveAddress.getOrElse(''))
  }
  if (equals('BTC', coin)) {
    let btcReceiveAddress
    if (type === ADDRESS_TYPES.LOCKBOX) {
      btcReceiveAddress = yield select(selectors.core.common.btc.getNextAvailableReceiveAddressLockbox(networks.btc, address, appState))
    } else {
      const defaultDerivation = yield select(selectors.core.common.btc.getAccountDefaultDerivation(address, appState))
      btcReceiveAddress = yield select(selectors.core.common.btc.getNextAvailableReceiveAddress(networks.btc, address, defaultDerivation, appState))
    }
    if (isEmpty(btcReceiveAddress.getOrElse(''))) {
      throw new Error('Could not generate return bitcoin receive address')
    }
    return btcReceiveAddress.getOrElse('')
  }
  throw new Error('Could not generate receive address')
}
