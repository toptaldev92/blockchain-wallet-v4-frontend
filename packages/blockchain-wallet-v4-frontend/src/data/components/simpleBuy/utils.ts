import { SwapAccountType } from '../swap/types'
import { SwapOrderDirectionType } from 'core/types'

export const getDirection = (
  from: SwapAccountType
): Exclude<SwapOrderDirectionType, 'TO_USERKEY' | 'ON_CHAIN'> => {
  switch (true) {
    case from.type === 'ACCOUNT':
      return 'FROM_USERKEY'
    default:
      return 'INTERNAL'
  }
}
