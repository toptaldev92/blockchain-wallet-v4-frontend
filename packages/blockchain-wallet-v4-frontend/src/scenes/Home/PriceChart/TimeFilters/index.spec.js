import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { testPropTypes } from 'utils/test.utils'
import { TimeFiltersContainer } from './index'

jest.mock('./template', () => 'template.na.js')
jest.mock('data', () => ({
  selectors: {
    modules: {
      profile: {
        isSilverOrAbove: jest.fn()
      }
    },
    components: {
      priceChart: {
        getTime: jest.fn()
      }
    }
  }
}))

describe('TimeFilters container', () => {
  it('renders correctly', () => {
    const baseProps = { time: '1month', isSilverOrAbove: false }
    const component = shallow(<TimeFiltersContainer {...baseProps} />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('should accept a mandatory string for prop time', () => {
    expect(
      testPropTypes(
        TimeFiltersContainer,
        'time',
        ['all', '1day', '1week', '1month', '1year'],
        false
      )
    ).toBeTruthy()
    expect(
      testPropTypes(
        TimeFiltersContainer,
        'time',
        [0, undefined, null, {}],
        true
      )
    ).toBeTruthy()
  })
})
