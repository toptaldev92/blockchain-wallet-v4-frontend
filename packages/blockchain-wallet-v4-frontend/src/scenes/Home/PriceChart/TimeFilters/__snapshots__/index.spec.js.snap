// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeFilters container renders correctly 1`] = `
<template
  handleClick={[Function]}
  time="1month"
/>
`;
